csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1985,66 +2034,146 @@ predicate isUnreachableInCall(Node n, DataFlowCall call) {
+ * improve dispatch.
  */
-class DataFlowType = Gvn::GvnType;
+class DataFlowType extends TDataFlowType {
+  Gvn::GvnType asGvnType() { this = TGvnDataFlowType(result) }
+
+  Callable asDelegate() { this = TDelegateDataFlowType(result) }
+
+  /**
+   * Gets an expression that creates a delegate of this type.
+   *
+   * For methods used as method groups in calls there can be multiple
+   * creations associated with the same type.
+   */
+  Expr getADelegateCreation() {
+    exists(Callable callable |
+      lambdaCreationExpr(result, callable) and
+      this = TDelegateDataFlowType(callable)
+    )
+  }
+
+  final string toString() {
+    result = this.asGvnType().toString()
+    or
+    result = this.asDelegate().toString()
+  }
+}
 
 /** Gets the type of `n` used for type pruning. */
-Gvn::GvnType getNodeType(Node n) { result = n.(NodeImpl).getDataFlowType() }
+DataFlowType getNodeType(Node n) {
+  result = n.(NodeImpl).getDataFlowType() and
+  not lambdaCreation(n, _, _) and
+  not delegateCreationStep(_, n)
+  or
+  exists(Node arg |
+    delegateCreationStep(arg, n) and
+    result = getNodeType(arg)
+  )
+  or
+  n.asExpr() = result.getADelegateCreation()
+}
 
 /** Gets a string representation of a `DataFlowType`. */
 string ppReprType(DataFlowType t) { result = t.toString() }
 
-private class DataFlowNullType extends DataFlowType {
+private class DataFlowNullType extends Gvn::GvnType {
   DataFlowNullType() { this = Gvn::getGlobalValueNumber(any(NullType nt)) }
 
   pragma[noinline]
-  predicate isConvertibleTo(DataFlowType t) {
+  predicate isConvertibleTo(Gvn::GvnType t) {
     defaultNullConversion(_, any(Type t0 | t = Gvn::getGlobalValueNumber(t0)))
   }
 }
 
-private class DataFlowUnknownType extends DataFlowType {
-  DataFlowUnknownType() { this = Gvn::getGlobalValueNumber(any(UnknownType ut)) }
+private class GvnUnknownType extends Gvn::GvnType {
+  GvnUnknownType() { this = Gvn::getGlobalValueNumber(any(UnknownType ut)) }
 }
 
-private predicate uselessTypebound(DataFlowType t) {
-  t instanceof DataFlowUnknownType or
-  t instanceof Gvn::TypeParameterGvnType
+pragma[nomagic]
+private predicate uselessTypebound(DataFlowType dt) {
+  dt.asGvnType() =
+    any(Gvn::GvnType t |
+      t instanceof GvnUnknownType or
+      t instanceof Gvn::TypeParameterGvnType
+    )
 }
 
 pragma[nomagic]
-predicate typeStrongerThan(DataFlowType t1, DataFlowType t2) {
+private predicate typeStrongerThanNonDelegate(DataFlowType t1, DataFlowType t2) {
   t1 != t2 and
-  t1 = getANonTypeParameterSubTypeRestricted(t2)
+  t1.asGvnType() = getANonTypeParameterSubTypeRestricted(t2.asGvnType())
   or
-  t1 instanceof RelevantDataFlowType and
+  t1.asGvnType() instanceof RelevantGvnType and
   not uselessTypebound(t1) and
   uselessTypebound(t2)
 }
 
+bindingset[t1, t2]
+predicate typeStrongerThan(DataFlowType t1, DataFlowType t2) {
+  typeStrongerThanNonDelegate(t1, t2)
+  or
+  exists(t1.asDelegate()) and
+  t1 != t2
```

Comment:
`typeStrongerThan(t1, t2)` should imply `compatibleTypes(t1, t2)`, which this disjunct doesn't.